This commit changes the create and tidy scripts to:

m0dular opened the pull request:
* Accept arguments instead of positional parameters in metrics_tidy
* Pipe the results of find to tar directly
* Changes path validation to ensure the target directory ends in one of
  the services we collect metrics for
* Use lowercase variables to we're not being yelled at

m0dular commented on files/create-metrics-archive:
```diff
@@ -1,29 +1,22 @@
-find . -type f -ctime -"$RETENTION_DAYS" -regex '\(.*gz$\|.*json$\)' > "$METRICS_OUTPUT_DIR.tmp"
-tar --create --gzip --file "$OUTPUT_FILE" --files-from "$METRICS_OUTPUT_DIR.tmp" --transform "s,^,/puppet-metrics-$DATETIME/,"
-rm "$METRICS_OUTPUT_DIR.tmp"
+find "$metrics_dir" -type f -ctime -"$ret_days" -a \( -name "*json" -o -name "*gz" \) | \
```
Here's an example of grouping the `-name` arguments instead of using a regex.

Also, we can take advantage of the fact that `tar` will remove the leading `/` from filenames in the `--transform` argument.  `${metrics_dir#/}` strips the leading slash, and replaces the match with `puppet-metrics`.

Lastly, piping the output of `find` to `tar` is better than creating a temp file.

tkishel replied:
Okay!

m0dular commented on files/metrics_tidy:
```diff
@@ -1,43 +1,33 @@
-METRICS_TYPE=$(basename "$METRICS_DIR")
+metrics_dir="${metrics_dir:-/opt/puppetlabs/puppet-metrics-collector/puppetserver}"
+ret_days="${ret_days:-90}"
+metrics_type="${metrics_dir##*/}"
```
Parameter expansion is preferred to an external command like `basename` because it is a shell builtin and is faster.

tkishel replied:
Yes, but this saves less than one second for a command that is not run interactively, at the expensive of readability (for those without advanced shell scripting skills) `##*/` is less self-documenting than `basename` which is common across programming languages.

```
real	0m0.002s
user	0m0.000s
sys	0m0.002s
```

```
real	0m0.003s
user	0m0.002s
sys	0m0.000s
```

But if you prefer, let's keep it.

m0dular replied:
Parameter expansion has been part of POSIX for a long time, and we should not be limiting ourselves based on what other may consider "readable" or "self-documenting," within reason of course.  I'll add a comment explaining what the line does.

tkishel replied:
Actually, we need to consider others who might want to contribute to the project (or users who want to understand what this script does on their system) but have less experience with shell scripting.

m0dular replied:
Is the comment explaining the line a good compromise?

tkishel replied:
Yes! 

Both the comment, and the `metrics_type` self-documenting variable name help avoid situations like this:

https://unix.stackexchange.com/questions/22387/how-do-0-and-0-work

tkishel commented on files/metrics_tidy:
```diff
@@ -1,43 +1,33 @@
+metrics_type="${metrics_dir##*/}"
 
 # Guard against deleting or archiving other files. 
+valid_paths=(puppetserver puppetdb orchestrator ace bolt activemq)
```
This creates a maintenance issue.
If you prefer to keep it, please document it in the manifests calling `metrics_tidy`

m0dular replied:
Added a comment in the manifests.

tkishel commented on files/create-metrics-archive:
```diff
@@ -1,29 +1,22 @@
+   esac
+   shift 2
 done
 
```
Maintain two space indents, please.

Also, I see now that `--metrics-directory` is redundant for a script named `create-metrics-archive`

Could we change this to `-d`  and `--directory`

m0dular replied:
I believe the indentation is correct, and `shellcheck` does not complain about it.  The lines after the condition ending in `)` and before the closing `;;` are indented to show they are part of that condition.  Using a real editor like `nvim` will auto-indent the lines like this.

I can change the parameter to `-d` and `--directory`.

tkishel replied:
"real" is inappropriate unless the other editors are imaginary.
Indenting using three spaces is unexpected.

m0dular replied:
Oh, I see what you're saying.  I don't know how that happened, since I do use two space indents by default.  Changed it to 2.

tkishel commented on files/metrics_tidy:
```diff
@@ -1,43 +1,33 @@
+   esac
+   shift 2
+done
 
```
Same comments as in `create-metrics-archive`

tkishel commented:
Thank you for this. Could you honor the use of $UPPERCASE for variables for readability, and long parameters (for command like `tar`) to improve self-documentation within the scripts?

tkishel commented:
The changes to `create-metrics-archive ` result in the expanded archive losing its timestamp:

````
[root@pe-201930-master ~]# ./create-metrics-archive 
tar: Removing leading `/' from member names
Created metrics archive: puppet-metrics-collector-2020.02.18.20.13.51.tar.gz

[root@pe-201930-master ~]# ll
total 388
-rwxr-xr-x. 1 root root    629 Feb 18 19:13 create-metrics-archive
-rw-r--r--. 1 root root 387043 Feb 18 20:22 puppet-metrics-collector-2020.02.18.20.13.51.tar.gz

[root@pe-201930-master ~]# tar -xzf puppet-metrics-collector*.tar.gz

[root@pe-201930-master ~]# ll
total 388
-rwxr-xr-x. 1 root root     629 Feb 18 19:13 create-metrics-archive
drwxr-xr-x. 10 root root    148 Feb 18 20:14 puppet-metrics
-rw-r--r--.  1 root root 387043 Feb 18 20:13 puppet-metrics-collector-2020.02.18.20.13.51.tar.gz
```

m0dular commented:
Made these changes: 

* Add timestamp to tarball in create script.  Thanks for pointing that out.
* Changed the parameter to `-d` and `--directory` in tidy and create scripts
* Added comments for the parameter expansion and valid paths.

I don't agree with uppercase variables in scripts.  These are generally system-wide variables, like:

```
$ echo ${!BASH*}

BASH BASHOPTS BASHPID BASH_ALIASES BASH_ARGC BASH_ARGV BASH_CMDS BASH_COMMAND BASH_COMPLETION_VERSINFO BASH_LINENO BASH_REMATCH BASH_SOURCE BASH_SUBSHELL BASH_VERSINFO BASH_VERSION
```
Lowercase variables in user scripts makes more sense to me.  I also think `tar czf` is standard usage and changing it to long options doesn't add much.

tkishel commented:
> I don't agree with uppercase variables in scripts. These are generally system-wide variables, like:
> Lowercase variables in user scripts makes more sense to me.

You are right, thanks for correcting me.

I've used uppercase for shell variables to differentiate them from other words (much like I used uppercase for reserved words like `SELECT * from projects` in SQL) and have always used long variable names with underscores to avoid collisions with ENV variables, but see the inherent risk.

>  I also think `tar czf` is standard usage and changing it to long options doesn't add much.

On this point, I still think these:

```
find "$dir" -type f -ctime -"$days" -and \( -name "*json" -or -name "*gz" \)

tar --create --gzip --file "$dir"
```

are easier for those less familiar to understand than:

```
find "$dir" -type f -ctime -"$days" -a \( -name "*json" -o -name "*gz" \)

tar czf "$dir"
```

But I'll cede this point for now.

tkishel commented:
One last item: your changes to metrics_tidy dropped deleting the files that have been archived:

```
xargs -a "$DIR.tmp" rm
```

:(

It deletes items older than days, archives the remainder, and deletes items it archives. Though as Jarret pointed out, since it's a separate command, it may do so even if the archiving fails?

m0dular commented:
I did think about the change around deleting files, and I think the only thing that would change is they'll be deleted during the next tidy run?  We do `find -delete` and then `find | tar`.

Could you invite Jarret to review this PR and comment on that?  I don't think I can since it's in your namespace.
